### ISSUE TITLE
VÉLØ Runtime Timing Audit — prove scoring depth and stage durations

### ISSUE THREAD

**Issue by elpresidentepiff** [elpresidentepiff/velo-oracle-prime#74]
# VÉLØ Runtime Timing Audit

## Problem

We do not currently know how much time VÉLØ spends on each scoring stage.

GitHub Actions only receives HTTP 202 from `/api/trigger/score-daily`; this confirms the background subprocess launched, not that full scoring completed deeply.

## Required output

Create:

data/runtime_timing_audit_YYYY_MM_DD.json

and optionally persist timing metadata to `pipeline_runs`.

## Instrument these stages

- bootstrap/env load
- preflight
- racecard load
- RP/PDF profile load
- PDF intel attachment
- normalization
- score_race_velo_prime total
- scoring duration per race
- scoring duration per runner
- SQPE model duration
- specialist model duration
- ensemble duration
- Spotlight parsing duration
- RPD-C tagging duration
- ProductRouter duration
- Candidate router duration
- Racing API shadow enrichment duration
- Supabase persistence duration
- Telegram duration
- local backup duration
- total runtime

## Required JSON shape

```json
{
  "date": "YYYY-MM-DD",
  "commit_sha": "...",
  "source": "cache|api|rp_profile",
  "total_runtime_sec": 0.0,
  "stages": [
    {
      "stage": "normalize",
      "duration_sec": 0.0,
      "races": 0,
      "runners": 0
    }
  ],
  "race_timings": [
    {
      "race_id": "...",
      "course": "...",
      "off_time": "...",
      "runners": 0,
      "score_race_velo_prime_sec": 0.0,
      "per_runner_avg_ms": 0.0,
      "pdf_intel_attached_count": 0,
      "spotlight_parsed_count": 0
    }
  ]
}
```

## Critical proof questions

1. Is VÉLØ genuinely analysing RP/PDF intelligence before scoring?
2. How many runners received `pdf_intel`?
3. How many runners received Spotlight parsing?
4. Does Spotlight happen before or after ranking?
5. Does any stage complete suspiciously fast?
6. Is scoring using cached/API data instead of RP profile data?

## Acceptance criteria

- Timing audit JSON created on every scoring run.
- Stage durations printed in final report.
- Total runtime visible in `pipeline_runs`.
- Per-race scoring timing visible.
- PDF intel attach count visible.
- Spotlight parse count visible.
- No live scoring logic changes.
- No model weight changes.

**Comment by elpresidentepiff** [comment]
## PR #75 merged — commit 947077b

**Phase 1 timing audit is live.** All required checks passed (test/lint/type-check).

### What landed
- `_RuntimeTimer` class with 7 stage marks: `preflight`, `racecard_load`, `normalize`, `pdf_intel_preload`, `persist`, `telegram`, `local_backup`
- Per-race timing: `score_race_velo_prime_sec`, `per_runner_avg_ms`, `pdf_intel_attached_count`, `spotlight_parsed_count`
- Writes `data/runtime_timing_audit_YYYY_MM_DD.json` on every run
- Stdout stage table in Railway logs

### Confirmed by instrumentation
Spotlight parsing happens **after** `score_race_velo_prime()` returns — it cannot affect `velo_prime_prob`, rank, or tier in the current flow.

### Phase 1 known limitations (Copilot review, resolved)

1. **Bootstrap not timed**: `_RuntimeTimer` instantiates after `_bootstrap_runtime()` — env load time excluded from `total_runtime_sec`.
2. **`racecard_load` boundary**: Stage includes Supabase client setup and pipeline_run open between preflight and actual `load_racecards()` call.
3. **`pipeline_runs` not patched**: `total_runtime_sec` not written to the Supabase `pipeline_runs` row — JSON file only.

### Phase 2 follow-up candidates
- Add `bootstrap` stage mark around `_bootstrap_runtime()`
- Add `pipeline_run_open` mark to isolate `load_racecards()` boundary cleanly
- Patch active `pipeline_runs` row with `total_runtime_sec` on successful completion
- Instrument inside `velo_prime_service.py`: SQPE / specialist / ensemble / TIE gate durations

**Comment by elpresidentepiff** [comment]
## First Runtime Timing Report — 2026-05-17

**Commit:** `947077b58416ef203c7ef99b39fc1f4962c97387`  
**Source:** cache (4 venues: Hamilton, Naas, Ripon, Stratford)  
**Mode:** dry-run (no Supabase persist, no Telegram)

---

### Totals

| Metric | Value |
|---|---|
| `total_runtime_sec` | **56.248s** |
| Races scored | 30 |
| Runners scored | 261 |
| `spotlight_runners_parsed` | **261 / 261 (100%)** |
| `pdf_intel_runners_attached` | **0** — no PDF files ingested for 2026-05-17 |

---

### Stage Breakdown

| Stage | Duration | % of total | Notes |
|---|---|---|---|
| `preflight` | 22.730s | **40.4%** | Enrichment cache load + Supabase health |
| `racecard_load` | 6.340s | 11.3% | Incl. heavy module imports + SB client init |
| `normalize` | 0.001s | 0.0% | — |
| `pdf_intel_preload` | 0.137s | 0.2% | 0/1 courses with PDF data |
| `persist`* | 26.658s | **47.4%** | Scoring loop + dry-run persist (see below) |
| `telegram` | 0.367s | 0.7% | — |
| `local_backup` | 0.014s | 0.0% | — |

\* Phase 1 limitation: the `persist` mark captures time since `pdf_intel_preload`, which in dry-run mode = **entire scoring loop** (no actual persist calls). A future `scoring_loop` mark would isolate this cleanly.

---

### Scoring Performance

| Metric | Value |
|---|---|
| Sum of `score_race_velo_prime_sec` | **21.779s** across 30 races |
| Avg per race | 726ms |
| Avg per runner (warm, races 2–30) | **78.1ms** |
| Avg per runner (global incl. cold-start) | 83.4ms |
| Loop overhead per race (RPDC/router/shadow/etc.) | **163ms** |

---

### Cold-Start Outlier — Hamilton 15:45

First race in the scoring loop is **4.0x slower** than warm races:

- `per_runner_avg_ms`: **309.3ms** (6 runners, 1.856s total)
- Warm avg: 78.1ms/runner

**Root cause:** First call to `score_race_velo_prime()` triggers lazy model loads (sklearn/xgboost deserialization) and JIT overhead. Every subsequent race benefits from warm model state.

**Implication:** The cold-start penalty is 1 race's worth of ~1.2s extra. Acceptable at current volume. If this becomes a latency concern, pre-warm models before the scoring loop.

---

### Slowest / Fastest Warm Races

| | Course | Time | Runners | `score_race_velo_prime_sec` | ms/runner |
|---|---|---|---|---|---|
| Slowest (warm) | Hamilton 18:15 | 15 runners | 1.212s | 80.8ms |
| Fastest | Stratford 15:05 | 5 runners | 0.383s | 76.6ms |

Timing is broadly linear with runner count. No pathological outliers beyond cold-start.

---

### What This Tells Us

1. **`preflight` (40%) is the dominant wall-clock cost** — enrichment caches, Supabase checks, heavy module imports. None of this is scoring logic. This is where to look first if Railway cron is slow to return.

2. **`score_race_velo_prime()` is not the bottleneck** — 21.8s for 30 races (261 runners) = 78ms/runner warm. That's fast enough. The per-runner cost is dominated by ensemble inference, not overhead.

3. **163ms/race of non-scoring loop overhead** (RPDC lookup, shadow enrichment, product router, candidate router, shadow ledger append) — this is real work, but not excessive.

4. **PDF intel = 0** for this date. This is expected (no PDF files ingested for 2026-05-17). On a live day with PDF files, `pdf_intel_runners_attached` will be >0. The preload stage (0.137s) confirms it checks and finds nothing — not broken.

5. **Spotlight = 100% coverage** — all 261 runners had Spotlight text. Confirmed it runs post-scoring and cannot affect `velo_prime_prob`.

---

### Phase 2 Targets (from this data)

- Add `bootstrap` stage mark around `_bootstrap_runtime()` to isolate it from `preflight`
- Add `pipeline_run_open` mark to separate SB client init from actual racecard load
- Add `scoring_loop` mark to isolate STEP 3 from the `persist` stage duration
- Instrument inside `velo_prime_service.py`: SQPE / specialist / ensemble / TIE gate durations
- Patch `pipeline_runs` row with `total_runtime_sec` on successful completion
- Model pre-warm before scoring loop to eliminate cold-start penalty (low priority)
